contracts/src/rollup/RollupCore.sol

```diff
@@ -400,35 +399,56 @@ abstract contract RollupCore is IRollupCore, PausableUpgradeable {
         {
             uint64 afterInboxPosition = assertion.afterState.globalState.getInboxPosition();
             uint64 prevInboxPosition = assertion.beforeState.globalState.getInboxPosition();
+
+            // there are 3 kinds of assertions that can be made. Assertions must be made when they fill the maximum number
+            // of blocks, or when they process all messages up to prev.nextInboxPosition. When they fill the max
+            // blocks, but dont manage to process all messages, we call this an "overflow" assertion.
+            // 1. ERRORED assertion
+            //    The machine finished in an ERRORED state. This can happen with processing any
+            //    messages, or moving the position in the message.
+            // 2. FINISHED assertion that did not overflow
+            //    The machine finished as normal, and fully processed all the messages up to prev.nextInboxPosition.
+            //    In this case the inbox position must equal prev.nextInboxPosition and position in message must be 0
+            // 3. FINISHED assertion that did overflow
+            //    The machine finished as normal, but didn't process all messages in the inbox.
+            //    The inbox position can be anything (within valid range) except the prev.nextInboxPosition,
+            //    the position in message can be anything except 0 if the inboxPosition == prev.inboxPosition.
+
+            //    All types of assertion must have inbox position in the range prev.inboxPosition <= x <= prev.nextInboxPosition
             require(afterInboxPosition >= prevInboxPosition, "INBOX_BACKWARDS");
-            if (assertion.afterState.machineStatus == MachineStatus.ERRORED) {
-                // the errored position must still be within the correct message bounds
+            require(afterInboxPosition <= assertion.beforeStateData.configData.nextInboxPosition, "INBOX_TOO_FAR");
+
+            // if the position in the message is > 0, then the afterInboxPosition cannot be the nextInboxPosition
+            // as this would be outside the range
+            if (assertion.afterState.globalState.getPositionInMessage() > 0) {
                 require(
-                    afterInboxPosition <= assertion.beforeStateData.configData.nextInboxPosition,
-                    "ERRORED_INBOX_TOO_FAR"
+                    afterInboxPosition != assertion.beforeStateData.configData.nextInboxPosition, "POSITION_TOO_FAR"
                 );
+            }
 
```

Comment:
for ERRORED state, we can also check if the posInMsg is backward

Reply:
Added check